File: solution/sliding-window/003-longest-substring-without-repeating-characters/solution.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function (s) {
    let left = 0;
    let maxLength = 0;

    const window = new Set();

    for (let right = 0; right < s.length; right++) {
        while (window.has(s[right])) {
            window.delete(s[left]);
            left++;
        }

        window.add(s[right]);

        maxLength = Math.max(maxLength, right - left + 1);
    }

    return maxLength;
};

File: solution/sliding-window/003-longest-substring-without-repeating-characters/solution.go
package slidingwindow

func lengthOfLongestSubstring(s string) int {
	left := 0
	maxLength := 0
	window := make(map[byte]bool)

	for right := 0; right < len(s); right++ {
		for window[s[right]] {
			window[s[left]] = false
			left++
		}

		window[s[right]] = true

		currentLength := right - left + 1
		if currentLength > maxLength {
			maxLength = currentLength
		}
	}

	return maxLength
}